zerver/actions/message_send.py

```diff
@@ -985,6 +988,56 @@ def do_send_messages(
                     )
                     send_request.automatic_new_visibility_policy = new_visibility_policy
 
+            # Set the visibility_policy of the users mentioned in the message
+            # to "FOLLOWED" if "automatically_follow_topics_where_mentioned" is "True".
+            active_users = send_request.active_user_ids - send_request.all_bot_user_ids
+
+            user_mentions = send_request.rendering_result.mentions_user_ids & active_users
+            user_wildcard_mentions = (
+                send_request.stream_wildcard_mention_user_ids
+                | send_request.topic_wildcard_mention_user_ids
+            ) & active_users
+            user_profiles_query_set: Set[UserProfile] = set()
+            user_profiles_query_set_wildcard_mentions: Set[UserProfile] = set()
+
+            if user_wildcard_mentions:
+                user_profiles_query_set_wildcard_mentions = set(
+                    UserProfile.objects.filter(
+                        id__in=user_wildcard_mentions,
+                        automatically_follow_topics_where_mentioned=True,
+                        wildcard_mentions_notify=True,
+                    )
+                )
+
+            if user_mentions:
+                user_profiles_query_set = set(
+                    UserProfile.objects.filter(
+                        id__in=user_mentions,
+                        automatically_follow_topics_where_mentioned=True,
+                    )
+                )
+            user_profiles = user_profiles_query_set | user_profiles_query_set_wildcard_mentions
+            if user_profiles:
+                user_topics_query_set = UserTopic.objects.filter(
+                    user_profile__in=user_profiles,
+                    stream_id=send_request.stream.id,
+                    topic_name__iexact=send_request.message.topic_name(),
+                    visibility_policy=UserTopic.VisibilityPolicy.FOLLOWED,
+                )
+                user_profiles_followed = {
+                    user_topic.user_profile for user_topic in user_topics_query_set
+                }
+
+                user_profiles_not_followed = list(user_profiles - user_profiles_followed)
+
+                if user_profiles_not_followed:
+                    bulk_do_set_user_topic_visibility_policy(
+                        user_profiles=user_profiles_not_followed,
+                        stream=send_request.stream,
+                        topic=send_request.message.topic_name(),
+                        visibility_policy=UserTopic.VisibilityPolicy.FOLLOWED,
+                    )
```

Comment:
So I think this algorithm is going to be too expensive specifically in the case of wildcard mentions in a very large organization. Let's drop from this PR the wildcard mention logic, and just have it apply to users who are directly mentioned (we can keep the wildcard test test, just with new assertions checking that it does not try to follow users).

I think like 95% of the value here is the personal mentions case, and like 95% of the risk is the wildcard mentions case, so let's just do the one and not the other.

Reply:
Sure, I have made changes so that this setting does not work for wildcard mentions.